### DISCUSSION TITLE
Try to add batch normalization in network.py, failed with dependency line.

### DISCUSSION THREAD

**Discussion by Franklalalala** [deepmodeling/deepmd-kit#1234] [Q&A]
I used high level api, tf.layers.batch_normalization(hidden, training=True, trainable=True)
They say, you need solve the dependency like:
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
with tf.control_dependencies(update_ops):
    train_op = optimizer.minimize(loss)
or, it'll not learn the gamma and belta.
I tried to add these lines in trainer.py, but failed with bugs.
I see there are lines about updatate variables.
https://github.com/deepmodeling/deepmd-kit/blob/159e45d248b0429844fb6a8cb3b3a201987c8d79/deepmd/train/trainer.py#L337-L354
Does that mean, I do not need to check the dependency anymore?
I'm new to tf, coundn't understand it.

I tried low API wrapper, always fail on line ↓
 with tf.control_dependencies([assign_moving_average(pop_mean, batch_mean, decay),
                                              assign_moving_average(pop_var, batch_var, decay)]):
                    return tf.identity(batch_mean), tf.identity(batch_var)
I don't know, how to slove this problem, or maybe, just leave it. ( It works without dependency line, with the high level API.)
Help!
